### PR TITLE
fix: prevent deadlock using torch multi worker

### PR DESCRIPTION
**Pull request by Jay-ju**
fix issue : https://github.com/lancedb/lance/issues/3640

### CODE REVIEW

File: python/python/tests/test_torch.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    yield str(dataset_path)
+
+    # Cleanup: remove entire test directory
+    if os.path.exists(test_dir):
+        for root, dirs, files in os.walk(test_dir, topdown=False):
+            for name in files:
+                os.remove(os.path.join(root, name))
+            for name in dirs:
+                os.rmdir(os.path.join(root, name))
+        os.rmdir(test_dir)
```
Comment (westonpace) [inline python/python/tests/test_torch.py:44]: Isn't this what `tmp_path_factory` already does?  Also, why not use `tmp_path`?

File: python/python/lance/torch/data.py [hunk 2]
```diff
@@ -375,3 +375,54 @@ def _blob_columns(self) -> List[str]:
...
+    def __init__(self, uri):
+        self.uri = uri
+        self._len = self._safe_preload()
+        self._ds = None  # Deferred initialization
+
+    def _safe_preload(self):
+        """Main-process safe metadata loading"""
+        ds = lance.dataset(self.uri)
+        length = ds.count_rows()
+        del ds  # Critical: release before spawning
```
Comment (westonpace) [inline python/python/lance/torch/data.py:390]: Why is it critical to `del ds` here?  What happens if you don't?

File: python/python/lance/torch/data.py [hunk 3]
```diff
@@ -375,3 +375,54 @@ def _blob_columns(self) -> List[str]:
...
+
+    def __getitem__(self, idx):
+        if self._ds is None:
+            # Worker-process initialization
+            import os
+
+            self._ds = lance.dataset(self.uri)
+            print(f"Worker {os.getpid()} initialized dataset")
+
+        return self._ds.take([idx]).to_pylist()[0]
```
Comment (westonpace) [inline python/python/lance/torch/data.py:404]: It's good to have a map-style dataset but we should clearly document this fact.  Users will be surprised by poor performance if they use a map-style dataset without any shuffling.

File: python/python/lance/torch/data.py [hunk 4]
```diff
@@ -375,3 +375,54 @@ def _blob_columns(self) -> List[str]:
...
+
+
+def get_safe_loader(dataset, batch_size=32, num_workers=4, **kwargs):
```
Comment (westonpace) [inline python/python/lance/torch/data.py:407]: Why are `batch_size` and `num_workers` special compared to the other parameters?

File: python/python/lance/torch/data.py [hunk 5]
```diff
@@ -375,3 +375,54 @@ def _blob_columns(self) -> List[str]:
                 logging.debug("Column %s is a Large Blob column", col)
                 blob_cols.append(col)
         return blob_cols
+
+
+class SafeLanceDataset(torch.utils.data.Dataset):
```
Comment (westonpace) [inline python/python/lance/torch/data.py:380]: What is the goal for this class?

Is it to introduce a map-style dataset?
Is it to avoid initialization of `self._ds`?

File: python/python/lance/torch/data.py [hunk 6]
```diff
@@ -375,3 +375,54 @@ def _blob_columns(self) -> List[str]:
...
+        """Main-process safe metadata loading"""
+        ds = lance.dataset(self.uri)
+        length = ds.count_rows()
+        del ds  # Critical: release before spawning
+        return length
+
+    def __len__(self):
+        return self._len
+
+    def __getitem__(self, idx):
```
Comment (westonpace) [inline python/python/lance/torch/data.py:396]: Implementing `__getitems__` will give a significant performance boost.

File: python/python/lance/torch/data.py [hunk 7]
```diff
@@ -375,3 +375,54 @@ def _blob_columns(self) -> List[str]:
...
+    # Approach 2: Explicit context creation
+    ctx = torch.multiprocessing.get_context("spawn")
+    return torch.utils.data.DataLoader(
+        dataset,
+        batch_size=batch_size,
+        num_workers=num_workers,
+        persistent_workers=True,
+        multiprocessing_context=ctx,  # Proper context object
+        # Critical for spawn compatibility:
+        pin_memory_device="cuda" if torch.cuda.is_available() else "",
```
Comment (westonpace) [inline python/python/lance/torch/data.py:426]: What is this line for?

File: python/python/lance/torch/data.py [hunk 8]
```diff
@@ -375,3 +375,54 @@ def _blob_columns(self) -> List[str]:
...
+    # (Must be in __main__ guard)
+    # torch.multiprocessing.set_start_method('spawn')
+
+    # Approach 2: Explicit context creation
+    ctx = torch.multiprocessing.get_context("spawn")
+    return torch.utils.data.DataLoader(
+        dataset,
+        batch_size=batch_size,
+        num_workers=num_workers,
+        persistent_workers=True,
```
Comment (westonpace) [inline python/python/lance/torch/data.py:423]: Why force this?

**Review by westonpace — CHANGES_REQUESTED** [review]
I like the idea of making the dataset more foolproof.

I'm not sure how these new classes are more discoverable than our existing documentation stating that users should use the `spawn` mode?

These methods need docstrings.

What if, for both this dataset and also `LanceDataset`, we do something like this...

```
import multiprocessing

class LanceDataset(torch.utils.data.IterableDataset):

    def __init__(
        self,
        ...
        *args,
        ...
        ignore_start_method=False,
    ):
        start_method = multiprocessing.get_start_method(allow_none=False)
        if start_method != "spawn" and not ignore_start_method:
            raise ValueError("lance datasets must use the 'spawn' start method.  Please call multiprocessing.set_start_method to change this.  If you are using a custom multiprocessing context you may set `ignore_start_method=True` to bypass this error")
```

File: python/python/lance/torch/data.py (hunk 2 above)
Comment (Jay-ju) [reply python/python/lance/torch/data.py:390]: Here, ds hopes to be reinitialized in the worker's process. Therefore, ds will be reassigned to None here.

File: python/python/lance/torch/data.py (hunk 3 above)
Comment (Jay-ju) [reply python/python/lance/torch/data.py:404]: The shuffling here is expected to be done in the sampler rather than in the Dataset. Therefore, no shuffle logic is provided here. The num_worker below is also for this reason.

File: python/python/lance/torch/data.py (hunk 4 above)
Comment (Jay-ju) [reply python/python/lance/torch/data.py:407]: Ditto. If you think the implementation here is relatively easy to use, I can also add it here. However, personally, I still hope that users can achieve the allocation of shuffle and num_workers by calling Dataset through the sampler and the torch framework. 
I can independently implement a RandomSampler or DistributeSampler with num_workers and batch_size.

File: python/python/lance/torch/data.py (hunk 5 above)
Comment (Jay-ju) [reply python/python/lance/torch/data.py:380]: Yes, both of these are the goals implemented by this class. Currently, there is only iterable_dataset.

File: python/python/lance/torch/data.py (hunk 7 above)
Comment (Jay-ju) [reply python/python/lance/torch/data.py:426]: delete it

File: python/python/lance/torch/data.py (hunk 8 above)
Comment (Jay-ju) [reply python/python/lance/torch/data.py:423]: This is not mandatory. I have made modifications.

File: python/python/lance/torch/data.py [hunk 9]
```diff
@@ -375,3 +375,76 @@ def _blob_columns(self) -> List[str]:
...
+        ds = lance.dataset(self.uri)
+        length = ds.count_rows()
+        del ds  # Critical: release before spawning
+        return length
+
+    def __len__(self):
+        return self._len
+
+    def __getitem__(self, idx):
+        return self.get_items([idx])[0]  # 复用批量接口
```
Comment (yanghua) [inline python/python/lance/torch/data.py:397]: You should remove the Chinese comment.

Comment (Jay-ju) [reply python/python/lance/torch/data.py]: resolved

**Review by westonpace — APPROVED** [review]
I'll merge this.  I think it is additive and gives users more options.

I am hoping to come up with something clearer in https://github.com/lancedb/lance/issues/3727 but I think this is a good step.
